%&context
\setuppapersize[letter][letter]
\setuplayout[topspace=0.75in, backspace=0.75in, height=middle, width=middle]
\setupwhitespace[medium]
\setuppagenumbering[location=footer]
\setupheader[state=none]
\usemodule[tikz]
\usetikzlibrary[graphs, graphdrawing, quotes]
\usegdlibrary[trees]
\usemodule[simplefonts]
\setmainfont[gentiumplus]

\starttext

\title{Design of a Selfing Simulator, {\sc SelfingSim}}

\startsection[title={Introduction}]
This document describes the design of {\sc SelfingSim},
a forward-in-time population genetics simulator with selfing
under pure-hermaphroditism, androdioecy, or gynodioecy.
\stopsection

\startsection[title={Design}]
{\sc SelfingSim} simulates a constant size population without
generation overlap under the infinite-alleles model of mutation.
One run of {\sc SelfingSim} can be broken into three distinct phases:
initialization, iteration, and termination.
Explanations of each phase are provided below.

\startsubsection[title={Initialization}]
At the beginning of a simulation, {\sc SelfingSim} initializes a population
based on users' input.
The first step is to assign \m {N} reproducing orgniams with \m {L} autosomal
loci to the population.
All organisms are hermaphroditic under pure-hermaphroditism,
but only \m {N_h} organisms are hermaprhoditic under andro- and gynodioecies.
The rest of the population are either males (androdioecy) or females (gynodioecy).

After initializing organisms, genotypes of all organisms are determined.
{\sc SelfingSim} currently supports four ways of initializing genotypes:
\quotation {\tt monomorphic}, \quotation {\tt unique}, \quotation {\tt count},
and \quotation {\tt frequency}.
All these methods work with a single locus, and no method is provided
to specify multi-locus genotypes or haplotypes.

The first two methods are straightfoward.
All genes are identical under \quotation {\tt monomorphic} or distinct under
\quotation {\tt unique}.
Under \quotation {\tt count}, a state of a gene is uniformly sampled from
\m {M} alleles.
Finally, \quotation {\tt frequency} again randomly draws a state of a gene
similar to \quotation {\tt count}, but a user can specify frequency of
each allele.

Preliminary simulations have found that starting a simulation
with \quotation {\tt monomorphic} scheme takes a long time to
approach mutation-drift equilibrium.
Mutation rate determines the rate of the approach.
On the other hand, starting a simulation with \quotation {\tt unique}
scheme allows to approach the equilibrium at much faster pace.
Genetic drift quickly eliminates excessive standing variation from
the population.
Based on these findings, \quotation {\tt unique} scheme has been used
throughout my simulations.
\stopsubsection

\startsubsection[title={Iteration}]
After initializing a population,
{\sc SelfignSim} starts simulating evolution of the population
for \m {T_{max}} generations.
Additionally, \m {T_B} generations of burn-in can precede the main
iterations.
The simulator performs the exactly same sequence of events
during burn-in and main iterations.
Therefore, running a simulation longer has the same effect on
the outcome of the simulation as running a shorter simulation with
burn-in period.

One generation in a simulation consists of two events: mutation and mating.
At the beginning of a generation/iteration, mutations are introduced
to the current member of the population at the locus-specific rates
\m {M_i} for locus \m {i}.
These rates are user-specified.
When a gene mutates, its new state has not been present throughout the entire
history of the population.

After mutation, the simulater performs model-specific mating to generate
offspring.
In all three modes, a user can specify composite parameters \m {s^*} and
\m {H}, which appear on the left-hand side of the following expressions,
or fundamental parameters such as \m {\tau} and \m {\tilde{s}} appearing
on the right-hand side.
The only exception is \m {N_h} under andro- and gynodioecies.
This parameter needs to be specified no matter if the rest of the parameters
are fundamental or composite.

Internally, if fundamental parameters are provided, those parameters are
converted to the corresponding composite parameters \m {s^*} and \m {H}
before starting a simulation.
Then, {\sc SelfingSim} only uses the composite parameters during a
simulation.

\startsubsubsection[title={Pure-Hermaphroditism}]
Under pure hermaphroditism, a selfing parent is randomly chosen from the
entire population with probability
\startformula
s^* = \frac{ \tilde {s} \tau }{ \tilde {s} \tau + 1 - \tilde {s} }.
\stopformula
Otherwise, two distinct parents are randomly chosen.
\stopsubsubsection

\startsubsubsection[title={Androdioecy}]
Under androdioecy, a selfing parent is randomly chosen from \m {N_h} hermaphroditic
parents with proability
\startformula
s^* = \frac{ \tilde {s} \tau }{ \tilde {s} \tau + 1 - \tilde {s} }.
\stopformula
Otherwise, two distinct parents, one male and one hermaphrodite, are randomly
chosen from \m {N - N_h} males and \m {N_h} hermaphrodites.

\stopsubsubsection

\startsubsubsection[title={Gynodioecy}]
Under gynodioecy, a selfing parent is randomly chosen from \m {N_h} hermaphroditic
parents with probability
\startformula
s^* = \frac{ \tau N_h a }{ \tau N_h a + N_h (1 - a) + N_f \sigma }.
\stopformula
Two distinct parents are chosen with probability \m { 1 - s^* }.
When there are two parents, with probability
\startformula
H = \frac{ N_h (1 - a) }{ N_h (1 - a) + N_f \sigma },
\stopformula
both parents are hermaphroditic.
Otherwise, one parent is hermaphroditic, and the other is female.
\stopsubsubsection

After parent(s) of an offspring are chonsen, the genotype of the offspring
is determined by the standard autosomal transimssion.
Furthermore, an identical recombination rate is applied to adjacent loci.
In addition to genetic transmission, {\sc SelfingSim} keeps track of the
number of generations since the last non-selfing event.

After mating and reproduction, the simulator can optionally save the state
of the entire population.
Stored information are state of genes and the number of generations since
the last outcrossing event.

Above two steps (or three with optional steps) are repeated for \m {T_{max}}
or \m {T_b + T_{max}} generations.

\stopsubsection

\startsubsection[title={Termination}, reference=sec:term]
During the termination phase of a simulation, the state of the
last generation is stored in file.
The information saved during this stage is identical to the ones stored
during the simulation (genotype and time since the last outcrossing event).

\stopsubsection

\stopsection

\startsection[title={Choosing Parent(s) under Gynodioecy}]
There are several ways to choose parents of an offspring under gynodioecy.
This section shows equivalence between those two parent-choosing methods.

In order to choose parent(s) under gynodioecy, three events have to occur
in sequence.
These events are a) the number of parents, b) sex of seed parent, and c)
survival of an offspring until its sexual maturity.
Two methods differ in whether the number of parents is decided before sex of seed parents or not.

Survival of an offspring is always determined the last. Let \m
{\{\tau'_i \le 0|i=\{1,2\}\}} be absolute fitness of uniparental or
biparental offspring.
Define \m {\tau_i = \tau'_i / \max(\tau'_1, \tau'_2)} as the rate of survival
relative to organisms with higher fitness.

\startsubsection[title={Sex before uniparental/biparental}]
One way to choose parent(s) is to determine the sex of the seed-parent before
deciding the number of parents.
The following decision tree shows a sequence of events starting
from a root.
Each edge is labeled by probability.

\starttikzpicture
\graph[tree layout, level distance=2cm, sibling distance=2cm]
{
    root -> {
        female [>"$\frac{N_f \sigma}{N_h + N_f \sigma}$"],
        hermaprhodite [>"$\frac{N_h}{N_h + N_f \sigma}$"]},
    female -> fb [as=biparent, >"$1$"],
    hermaprhodite -> {
        hu [as=uniparent, >"$a$"],
        hb [as=biparent, >"$1-a$"]},
    fb -> {fba [as=alive, >"$\tau_2$"], fbd [as=dead, >"$1-\tau_2$"]},
    hu-> {hua [as=alive, >"$\tau_1$"], hud [as=dead, >"$1-\tau_1$"]},
    hb -> {hba [as=alive, >"$\tau_2$"], hbd [as=dead, >"$1-\tau_2$"]};
};
\stoptikzpicture

In the end, there are six possible outcomes, of which only three lead
to viable offspring.
The following table summarizes joint probabilities of all events and
joint probability of the sex of seed-parent and the number of parents
conditional on survival of offspring.

\starttabulate[|c|c|c|m|m|]
\HL
\NC Sex \NC Number \NC Survival
\NC {\rm Probability} \NC {\rm Probility | Alive} \NR
\HL
\NC Female \NC 2 \NC Alive
\NC \frac{N_f \sigma}{N_h + N_f \sigma } \tau_2
\NC \frac{N_f \sigma \tau_2}
{N_h a \tau_1 + N_h (1 - a) \tau_2 + N_f \sigma \tau_2}  \NR

\NC Female \NC 2 \NC Dead
\NC \frac{N_f \sigma}{N_h + N_f \sigma} (1 - \tau_2) \NC \NR

\NC Hermaphrodite \NC 1 \NC Alive
\NC \frac{N_h}{N_h + N_f \sigma} a \tau_1
\NC \frac{N_h a \tau_1}
{N_h a \tau_1 + N_h (1 - a) \tau_2 + N_f \sigma \tau_2} \NR

\NC Hermaphrodite \NC 1 \NC Dead
\NC \frac{N_h}{N_h + \sigma N_f} a (1 - \tau_1) \NC \NR

\NC Hermaphrodite \NC 2 \NC Alive
\NC \frac{N_h}{N_h + \sigma N_f} (1 - a) \tau_2
\NC \frac{N_h (1 - a) \tau_2}
{N_h a \tau_1 + N_h (1 - a) \tau_2 + N_f \sigma \tau_2} \NR

\NC Hermaprhodite \NC 2 \NC Dead
\NC \frac{N_h}{N_h + N_f \sigma} (1 - a) (1 - \tau_2) \NC \NR
\HL
\stoptabulate
\stopsubsection

\startsubsection[title={Uniparental/Biparental before Sex}]

Another way to choose parent(s) is to decide the number of parents before
deciding the sex of seed-parent as depicted below.

\starttikzpicture
\graph[tree layout, level distance=2cm, sibling distance=2cm]
{
    root -> {
        uniparent [>"$\frac{N_h a}{N_h + N_f \sigma}$"],
        biparent [>"$\frac{N_h (1 - a) + N_f \sigma}{N_h + N_f \sigma}$"]},
    uniparent -> hu [as=hermaprhodite, >"$1$"],
    biparent -> {
        fb [as=female, >"$\frac{N_h(1-a)}{N_h(1-a) + N_f \sigma}$"],
        hb [as=hermaprhodite, >"$\frac{N_f \sigma}{N_h(1-a) + N_f \sigma}$"]},
    hu -> {hua [as=alive, >"$\tau_1$"], hud [as=dead, >"$1-\tau_1$"]},
    fb -> {fba [as=alive, >"$\tau_2$"]], fbd [as=dead, >"$1-\tau_2$"]},
    hb -> {hba [as=alive, >"$\tau_2$"], hbd [as=dead, >"$1-\tau_2$"]};
};
\stoptikzpicture

Again, the following table surmarizes joint and conditonal probabilities of
events.

\starttabulate[|c|c|c|m|m|]
\HL
\NC Number \NC Sex \NC Survival
\NC {\rm Probability} \NC {\rm Probability | Alive} \NR
\HL

\NC 1 \NC Hermaphrodite \NC Alive
\NC \frac{N_h a}{N_h + N_f \sigma} \tau_1
\NC \frac{N_h a \tau_1}
{N_h a \tau_1 + N_h (1 - a) \tau_2 + N_f \sigma \tau_2} \NR

\NC 1 \NC Hermaphrodite \NC Dead
\NC \frac{N_h a}{N_h + N_f \sigma} (1 - \tau_1) \NC \NR

\NC 2 \NC Female \NC Alive
\NC \frac{N_f \sigma}{N_h + N_f \sigma} \tau_2
\NC \frac{N_f \sigma \tau_2}
{N_h a \tau_1 + N_h (1 - a) \tau_2 + N_f \sigma \tau_2} \NR

\NC 2 \NC Female \NC Dead
\NC \frac{N_f \sigma}{N_h + N_f \sigma} (1 - \tau_2) \NC \NR

\NC 2 \NC Hermaphrodite \NC Alive
\NC \frac{N_h (1 - a)}{N_h + N_f \sigma} \tau_2
\NC \frac{N_h (1 - a) \tau_2}
{N_h a \tau_1 + N_h (1 - a) \tau_2 + N_f \sigma \tau_2} \NR

\NC 2 \NC Hermaphrodite \NC Dead
\NC \frac{N_h (1 - a)}{N_h + N_f \sigma} (1 - \tau_2) \NC \NR
\HL
\stoptabulate

\stopsubsection

\startsubsection[title={Equivalence}]
Comparison of the last column from both tables show that both methods
lead to outcomes with identical probabilities.
\stopsubsection

\stopsection

\stoptext
