# -*- mode: python; coding: utf-8; -*-

# infinite_sites.py - A collection of evolutionary operators specific
# for the infinite sites model.

# Copyright (C) 2013 Seiji Kumagai

# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice (including the next
# paragraph) shall be included in all copies or substantial portions of the
# Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.


import simuPOP as simu
from collections import deque

def get_mating_operator(r_rate, loci, allele_length, weight, size):
    """
    Construct partially selfing mating operation under the infinite sites model.

    A fraction, 0 <= weight <= 1, of offspring is generated by selfing, and others are
    generated by outcrossing.  In this model, there is no specific sex so that any
    individual can mate with any other individuals in a population.
    Furthermore, a parent can participate in both selfing and outcrossing.
    """
    # Index of sites, after which recombinations happen.
    rec_loci = [allele_length * i - 1 for i in range(1, loci + 1)]
    selfing = simu.SelfMating(ops = [sim.Recombinator(rates = recomb_rate,
                                                      loci = rec_loci),
                                     cf.MySelfingTagger()],
                              weieght = weight)

    outcross = simu.HomoMating(chooser = simu.PyParentsChooser(generator = cf.pickTwoParents),
                               generator = simu.OffSpringGenerator(
                                   ops = [simu.Recombinator(rates = r_rate,
                                                            loci = rec_loci)
                                          cf.MyOutcrossingTagger()]),
                               weight = 1.0 - weight)

    return simu.HeteroMating(matingSchemes = [selfing, outcross],
                               subPopSize = size)


def get_mutation_operator(m_rate, loci, allele_length, nrep, burnin):
    class MyMutator(simu.PyOperator):
        """
        A mutation operator representing the infinite sites model.

        A new mutation occurs at a distinct site from any previous mutated sites.
        """
        def __init__(self, m_rate, loci, allele_length, nrep, burnin):
            self.available = {r: [deque(range(allele_length)
                                        for i in range(loci))]
                              for r in range(rep)}
            self.m_rate = m_rate
            self.loci = loci
            self.allele_length = allele_length
            self.burnin = burnin

            super(MyMutator, self).__init__(func = self.mutate)


        def mutate(self, pop):
            """Add mutations to organisms."""
            m_rate = self.m_rate
            available = self.available

            rng = simu.getRNG()

            dvars = pop.dvars()
            rep = dvars.rep
            gen = dvars.gen - self.burnin

            for i, ind = enumerate(pop.individuals()):
                for locus in range(loci):
                    for ploidy in range(2):
                        if rng.randomUniform() < m_rate:
                            try:
                                idx = available[rep][locus].pop()
                            except IndexError:
                                if self.reclaim(pop, rep, locus):
                                    idx = self.available[rep][locus].pop()
                                else:
                                    sys.stderr.write(
                                        'rep={}, gen={}: available sites exhausted.'.
                                        format(rep, gen - burnin))
                                    return False

                            idx += locus * allele_length
                            # A site is represented by 1 bit.  An
                            # ancestral state is always represented by
                            # 0, and mutated state is therefore always
                            # 1.
                            ind.setAllele(1, idx, ploidy = ploidy)
            return True


        def reclaim(self, pop, rep, locus):
            """
            Scan and reuse monomorphic sites.

            A new mutation uses up one available site/locus.  After enough mutations
            appear in a population, a simulation exhausts all sites/loci.  However,
            some of such mutations may be either extinct from or fixed in a current
            population.  Extinct mutations will never come back to the population, and
            fixed mutations likewise never returns to ancestral states.  Therefore,
            those sites occupied by such mutations can be freed and reused.  This method
            scans all sites to find monomorphic sites.  Once the monomorphic sites are
            identified, those sites are re-initialized and registered to a list of all
            available sites.
            """
            allele_length = self.allele_length
            start = locus * allele_length
            stop = (locus + 1) * allele_length
            sites = list(range(start, stop))

            allele_states = areAllelesMonomorphic(pop, sites)
            available = [i for i, state in enumerate(allelee_states) if state == True]

            if len(available) == 0:
                # Even after scanning all sites, there is no unused site.
                return False

            # Re-initialize newly freed sites by setting their values 1.
            for ind in pop.individuals():
                for site in available:
                    ind.setAllele(0, site, ploidy = 0)
                    ind.setAllele(0, site, ploidy = 1)

            self.available[rep][locus] = deque(available)


        return MyMutator(m_rate = m_rate,
                         loci = loci,
                         allele_length = allele_length,
                         nrep = nrep,
                         burnin = burnin)


def get_output_operator():
    """Output genetic information of a population."""

    class MyWriter(simu.PyOperator):
        """A class handling output of genetic information of the entire population."""

        def __init__(self,
                     num_ind,
                     num_gen,
                     num_rep,
                     m_rate,
                     r_rate,
                     s_rate,
                     loci,
                     burnin,
                     seed,
                     output,
                     allele_length):
            # self.num_ind = num_ind
            # self.num_gen = num_gen
            # self.num_rep = num_rep
            # self.m_rate = m_rate
            # self.r_rate = r_rate
            # self.s_rate = s_rate
            self.loci = loci
            self.burnin = burnin
            self.output = output

            self.data = ['infinite sites',
                         num_ind,
                         num_gen,
                         num_rep,
                         m_rate,
                         r_rate,
                         s_rate,
                         loci,
                         burnin,
                         simu.getRNG().seed()]

            # Allele length is the only parameter that will not be
            # printed.  This variable controls the assignment of
            # sites, which can hold polymorphic sites, and it is
            # there for strictly an implementation reason (albeit user
            # configurable).
            self.allele_length

            self.header = ['mutation model',
                           'number of individuals',
                           'number of generations',
                           'number of replicates',
                           'number of loci',
                           'mutation rate',
                           'recombination rate',
                           'selfing rate',
                           'number of burnin generations',
                           'random number seed',
                           'replicate',
                           'generation',
                           'individual',
                           'chromosome'] + ['locus {}'.format(i) for i in range(loci)]

            with open(output, 'w') as f:
                writer = csv.DictWriter(f, self.header)
                writer.writeheader()

            super(MyWriter, self).__init__(func = self.write)

        def write(self, pop):
            output = self.output
            # In order to keep output file structure simple, all
            # information regarding to simulation such as model
            # parameters are included into each row.  This obviously
            # caused repetition of simulation-wide parameters many
            # times and excessive use of storage space.  However, I
            # consider an upside, the simplicity of the output file
            # structure, is well worth the cost.

            header = self.header
            dat = self.data

            with open(output, 'a') as f:
                dvars = pop.dvars()
                rep = dvars.rep
                gen = dvars.gen - self.burnin

                writer = cvs.DictWriter(f, data.keys())


                # scan genotype of all individuals to identify
                # polymorphic sites.

                # convert to carray to list because I want to use
                # slice
                loci = self.loci
                allele_length = self.allele_length
                raw_geno = list(pop.genotype())
                for locus in range(loci):
                    for site in range(allele_length):
                        idx = locus * site
                        poly_sites = [i for i in range(allele_length)
                                      if len(set(raw_geno[site::loci * allele_length])) > 1]

                for idx, ind in enumerate(pop.individuals()):
                    for ploidy in range(2):
                        geno = ind.genotype(ploidy = ploidy)
                        for locus in range(self.loci):
                        sites = [geno[i] for i in poly_sites]


                        writer.writerow({key: value for key, value in
                                         zip(header,
                                             data + [rep, gen, idx, ploidy] + geno)})

            return True
